### ISSUE TITLE
CCACHE_VERIFY?

### ISSUE THREAD

**Issue by peterwaller-arm** [ccache/ccache#950]
* I have what I believe is a deterministic compiler (assumption).
* I have encountered a case where `CCACHE_DISABLE=1` produces a different binary than with ccache enabled. (The runtime behaviour of the compiled artefact is drastically different with/without cache).
* I'm worried that my ccache is contaminated or ccache is buggy.
* I want to triage this; so I can report a bug to ccache if appropriate, or identify the root cause.
  * I'd prefer not to throw away my cache.
  * I would like to build trust in my cache.
* The first thing I'm interested to know is which binaries have diverged on recompilation.

Can ccache help with this problem? I searched the manual for 'verify' and had a pretty thorough look at the configurables but don't see it. I am imagining `CCACHE_VERIFY=1` would, on a cache hit, recompile the program and report when the output is different from the content of the cache hit.

Obviously, I can compile twice (with/without `CCACHE_DISABLE`) and compare checksums but this seems useful and reassuring functionality to have immediate access to. ccache could also help by identifying the erroneous cache entries so they can be deleted.

Other things I have considered, independently of CCACHE_VERIFY:

`CCACHE_RECACHE` would overwrite a divergent binary, losing precious information about the miscache/miscompile.
`CCACHE_DEBUG` says that it would provide information on "accidental misses", but I (seemingly) have an "accidental hit", so I haven't looked deeply at that yet.
`CCACHE_COMPILERCHECK` is left at its default value of mtime. I have thought about this and can see how it could cause the issue I'm experiencing but don't currently think it is applicable in my case.
I haven't configured anything else other than `CCACHE_FILECLONE`, `CCACHE_MAXSIZE`, `CCACHE_DIR` and `CCACHE_COMPILER`.

**Comment by jrosdahl** [comment]
> I have what I believe is a deterministic compiler (assumption).

You didn't say which ccache version you're using, but if it is (or was) 4.4 or 4.4.1 then there is a [known nasty bug](https://github.com/ccache/ccache/issues/935) that can produce false positive cache hits.

> Can ccache help with this problem

There is nothing available other than the options you already have found.

>  I am imagining `CCACHE_VERIFY=1` would, on a cache hit, recompile the program and report when the output is different from the content of the cache hit.

That sounds like a good idea.

**Comment by peterwaller-arm** [comment]
> You didn't say which ccache version you're using, but if it is (or was) 4.4 or 4.4.1

Good point! I was running a master branch of ccache, and it was at a commit shortly after the regression was introduced, so that probably explains it. Having a CCACHE_VERIFY option might make it easier for people to find this sort of bug and observe compiler non-determinism.
